esp-hal/src/system.rs

```diff
@@ -560,232 +585,200 @@ impl PeripheralClockControl {
 
 #[cfg(any(esp32c6, esp32h2))]
 impl PeripheralClockControl {
-    /// Enables and resets the given peripheral
-    pub(crate) fn enable(peripheral: Peripheral) {
-        let system = unsafe { &*SYSTEM::PTR };
+    pub(crate) fn enable_forced(peripheral: Peripheral, enable: bool, force: bool) -> bool {
+        debug!("Enable {:?} {} {}", peripheral, enable, force);
+        critical_section::with(|_| {
+            if !force {
+                if enable {
+                    let prev =
+                        PERIPHERAL_REF_COUNT[peripheral as usize].fetch_add(1, Ordering::Relaxed);
+                    if prev > 0 {
+                        return false;
+                    }
+                } else {
+                    let prev =
+                        PERIPHERAL_REF_COUNT[peripheral as usize].fetch_sub(1, Ordering::Relaxed);
+                    if prev > 1 {
+                        return false;
+                    }
+                };
+            } else if !enable {
+                PERIPHERAL_REF_COUNT[peripheral as usize].store(0, Ordering::Relaxed);
```

Comment:
I'm not sure what use case this is for but it could lead to an underflow if/when `fetch_sub` is called later.

Reply:
The only case when it's forced disabled is in `esp_hal::init` - so it should be okay (given this is internal API only).
Would be nice if there was a saturating `fetch_add`/`fetch_sub` - on the other hand if it ever wraps it would be a bug in esp-hal

Reply:
In that case, instead of calling `store`, perhaps use `swap` and assert that the previous value is also zero.
In the `fetch_sub` case, you can assert that the prev value isn't zero.

At least if there is a bug in the hal, it'd be caught rather than subtly break